doc/std/D2175.md

```diff
@@ -0,0 +1,3182 @@
+  // This will communicate the request to stop via the std::stop_token
+  // passed to the thread's entry-point function and this will interrupt
+  // the cv.wait_for() call and cause the thread to exit promptly.
+}
+```
+
+However, the `stop_token` abstraction is more general than `std::jthread`
+and can also be used for cancellation of asynchronous code in much the same
+way that `CancellationToken` is used in `folly::coro` and in the .NET Framework.
+
+A general purpose cancellation mechanism for sender/receiver would ideally integrate
+with the `std::stop_token` abstraction. Although it is worth noting that the design
+of `std::stop_token`, like `folly::CancellationToken`, requires a heap-allocation
+and reference-counting, which is not strictly necessary for structured concurrency
+use-cases.
+
+# Proposed Changes
+
+This section will describe the facilities being proposed by this paper.
+The following section, 'Design Discussion' will discuss some of the design
+considerations relating to this proposal.
+
+## Add `stoppable_token` and related concepts
+
+Add the following concept definition to the `<concepts>` header:
+```c++
+namespace std
+{
+  template<template<typename> class>
+  struct __check_type_alias_exists;  // exposition-only
+
+  template<typename T>
+  concept stoppable_token =
+    copy_constructible<T> &&
+    move_constructible<T> &&
+    is_nothrow_copy_constructible_v<T> &&
+    is_nothrow_move_constructible_v<T> &&
+    equality_comparable<T> &&
+    requires (const T& token) {
+      { token.stop_requested() } noexcept -> boolean-testable;
+      { token.stop_possible() } noexcept -> boolean-testable;
+      typename __check_type_alias_exists<T::template callback_type>;
+    };
+
+  template<typename T, typename CB, typename Initializer = CB>
+  concept stoppable_token_for =
+    stoppable_token<T> &&
+    invocable<CB> &&
+    requires {
+      typename T::template callback_type<CB>;
+    } &&
+    constructible_from<CB, Initializer> &&
+    constructible_from<typename T::template callback_type<CB>, T, Initializer> &&
+    constructible_from<typename T::template callback_type<CB>, T&, Initializer> &&
+    constructible_from<typename T::template callback_type<CB>, const T, Initializer> &&
+    constructible_from<typename T::template callback_type<CB>, const T&, Initializer>;
+
+  template<typename T>
+  concept unstoppable_token =
+    stoppable_token<T> &&
+    requires {
+      { T::stop_possible() } -> boolean-testable;
+    } &&
+    !T::stop_possible();
+}
+```
+
+The `stoppable_token` concept checks for the basic interface of a "stop token"
+which is copyable and allows polling to see if stop has been requested and also
+whether a stop request is possible.
+
+It also provides an associated nested template-type-alias, `T::callback_type<CB>`,
+that identifies the stop-callback type to use to register a callback to be executed
+if a stop-request is ever made on a `stoppable_token` of type, `T`.
```

Comment:
did we consider replacing this nested type alias with an algorithm?

a 'tag_invoke' based CPO named `make_callback()` and a trait to derive the return type would make it possible to support different callback types based on the arguments (like the invocable argument)

This also means that the existing `stop_token` does not need a change to its public interface in order to match the `stoppable_token` concept

Reply:
Yes, I had been considering something like this as well.
I'm not entirely sure why I ended up going with the type-alias, but it was probably because that's what we've implemented in libunifex.

I had also been wondering how the factory-function would accommodate being able to specify the callback type to store separate from the argument used to initialise it.
e.g. I can do: `ST::callback_type<C>{st, X{}}` and this will initialise C with the rvalue-reference to X.

I guess it could be `make_callback<C>(st, X{})` but you'd also want to be able to do something like `make_callback(st, C{})`.

It's probably worth exploring further.